// weapp-tailwindcss/dist/chunk-3CBQYS3U.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkLSSLYD6Bjs = require('./chunk-LSSLYD6B.js');


var _chunkEHEK77ROjs = require('./chunk-EHEK77RO.js');


var _chunkG2TGR4BFjs = require('./chunk-G2TGR4BF.js');


var _chunkO335YLYHjs = require('./chunk-O335YLYH.js');

// src/bundlers/vite/index.ts
var debug = _chunkLSSLYD6Bjs.createDebug.call(void 0, );
function UnifiedViteWeappTailwindcssPlugin(options = {}) {
  const opts = _chunkG2TGR4BFjs.getCompilerContext.call(void 0, options);
  const {
    disabled,
    onEnd,
    onLoad,
    onStart,
    onUpdate,
    templateHandler,
    styleHandler,
    jsHandler,
    mainCssChunkMatcher,
    appType,
    setMangleRuntimeSet,
    cache,
    twPatcher
  } = opts;
  if (disabled) {
    return;
  }
  twPatcher.patch();
  onLoad();
  return {
    name: _chunkEHEK77ROjs.vitePluginName,
    enforce: "post",
    async generateBundle(_opt, bundle) {
      debug("start");
      onStart();
      const entries = Object.entries(bundle);
      const groupedEntries = _chunkO335YLYHjs.getGroupedEntries.call(void 0, entries, opts);
      const runtimeSet = await twPatcher.getClassSet();
      setMangleRuntimeSet(runtimeSet);
      debug("get runtimeSet, class count: %d", runtimeSet.size);
      if (Array.isArray(groupedEntries.html)) {
        let noCachedCount = 0;
        for (const element of groupedEntries.html) {
          const [file, originalSource] = element;
          const oldVal = originalSource.source.toString();
          const hash = cache.computeHash(oldVal);
          cache.calcHashValueChanged(file, hash);
          await cache.process(
            file,
            () => {
              const source = cache.get(file);
              if (source) {
                originalSource.source = source;
                debug("html cache hit: %s", file);
              } else {
                return false;
              }
            },
            async () => {
              originalSource.source = await templateHandler(oldVal, {
                runtimeSet
              });
              onUpdate(file, oldVal, originalSource.source);
              debug("html handle: %s", file);
              noCachedCount++;
              return {
                key: file,
                source: originalSource.source
              };
            }
          );
        }
        debug("html handle finish, total: %d, no-cached: %d", groupedEntries.html.length, noCachedCount);
      }
      if (Array.isArray(groupedEntries.js)) {
        let noCachedCount = 0;
        for (const element of groupedEntries.js) {
          const [file, originalSource] = element;
          const rawSource = originalSource.code;
          const hash = cache.computeHash(rawSource);
          cache.calcHashValueChanged(file, hash);
          await cache.process(
            file,
            () => {
              const source = cache.get(file);
              if (source) {
                originalSource.code = source;
                debug("js cache hit: %s", file);
              } else {
                return false;
              }
            },
            async () => {
              const mapFilename = `${file}.map`;
              const hasMap = Boolean(bundle[mapFilename]);
              const { code, map } = await jsHandler(rawSource, runtimeSet, {
                generateMap: hasMap
              });
              originalSource.code = code;
              onUpdate(file, rawSource, code);
              debug("js handle: %s", file);
              noCachedCount++;
              if (hasMap && map) {
                ;
                bundle[mapFilename].source = map.toString();
              }
              return {
                key: file,
                source: code
              };
            }
          );
        }
        debug("js handle finish, total: %d, no-cached: %d", groupedEntries.js.length, noCachedCount);
      }
      if (Array.isArray(groupedEntries.css)) {
        let noCachedCount = 0;
        for (const element of groupedEntries.css) {
          const [file, originalSource] = element;
          const rawSource = originalSource.source.toString();
          const hash = cache.computeHash(rawSource);
          cache.calcHashValueChanged(file, hash);
          await cache.process(
            file,
            () => {
              const source = cache.get(file);
              if (source) {
                originalSource.source = source;
                debug("css cache hit: %s", file);
              } else {
                return false;
              }
            },
            async () => {
              const { css } = await styleHandler(rawSource, {
                isMainChunk: mainCssChunkMatcher(originalSource.fileName, appType)
              });
              originalSource.source = css;
              onUpdate(file, rawSource, css);
              debug("css handle: %s", file);
              noCachedCount++;
              return {
                key: file,
                source: css
              };
            }
          );
        }
        debug("css handle finish, total: %d, no-cached: %d", groupedEntries.css.length, noCachedCount);
      }
      onEnd();
      debug("end");
    }
  };
}



exports.UnifiedViteWeappTailwindcssPlugin = UnifiedViteWeappTailwindcssPlugin;
