// weapp-tailwindcss/dist/chunk-XWGMPET3.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkO335YLYHjs = require('./chunk-O335YLYH.js');

// src/defaults.ts
var _escape = require('@weapp-core/escape');
var _localpkg = require('local-pkg');
function getDefaultOptions() {
  return {
    /**
     * wxss 微信小程序
     * acss 支付宝小程序
     * jxss 京东小程序
     * ttss 头条小程序
     * qss QQ小程序
     * css 最正常的样式文件
     * tyss 涂鸦小程序
     */
    cssMatcher: (file) => /.+\.(?:wx|ac|jx|tt|q|c|ty)ss$/.test(file),
    /**
     * wxml 微信小程序
     * axml 支付宝小程序
     * jxml 京东小程序
     * ksml 快手小程序
     * ttml 头条小程序
     * qml QQ小程序
     * tyml 涂鸦小程序
     * xhsml 小红书小程序
     * swan 百度小程序
     */
    htmlMatcher: (file) => /.+\.(?:(?:wx|ax|jx|ks|tt|q|ty|xhs)ml|swan)$/.test(file),
    jsMatcher: (file) => {
      if (file.includes("node_modules")) {
        return false;
      }
      return /.+\.[cm]?js?$/.test(file);
    },
    mainCssChunkMatcher: (file, appType) => {
      switch (appType) {
        case "uni-app": {
          return file.startsWith("common/main");
        }
        case "uni-app-vite": {
          return file.startsWith("app") || file.startsWith("common/main");
        }
        case "mpx": {
          return file.startsWith("app");
        }
        case "taro": {
          return file.startsWith("app");
        }
        case "remax": {
          return file.startsWith("app");
        }
        case "rax": {
          return file.startsWith("bundle");
        }
        case "native": {
          return file.startsWith("app");
        }
        case "kbone": {
          return /^(?:common\/)?miniprogram-app/.test(file);
        }
        default: {
          return true;
        }
      }
    },
    wxsMatcher: () => {
      return false;
    },
    // https://tailwindcss.com/docs/preflight#border-styles-are-reset-globally
    cssPreflight: {
      "box-sizing": "border-box",
      "border-width": "0",
      "border-style": "solid",
      "border-color": "currentColor"
    },
    disabled: false,
    customRuleCallback: _chunkO335YLYHjs.noop,
    onLoad: _chunkO335YLYHjs.noop,
    onStart: _chunkO335YLYHjs.noop,
    onEnd: _chunkO335YLYHjs.noop,
    onUpdate: _chunkO335YLYHjs.noop,
    customAttributes: {},
    customReplaceDictionary: _escape.MappingChars2String,
    appType: void 0,
    arbitraryValues: {
      allowDoubleQuotes: false
    },
    cssChildCombinatorReplaceValue: ["view", "text"],
    inlineWxs: false,
    injectAdditionalCssVarScope: false,
    jsPreserveClass: (keyword) => {
      if (keyword === "*") {
        return true;
      }
      return false;
    },
    disabledDefaultTemplateHandler: false,
    cssSelectorReplacement: {
      root: "page",
      universal: ["view", "text"]
    },
    babelParserOptions: {
      sourceType: "unambiguous"
    },
    postcssOptions: {},
    cssRemoveHoverPseudoClass: true,
    ignoreCallExpressionIdentifiers: _localpkg.isPackageExists.call(void 0, "@weapp-tailwindcss/merge") ? ["twMerge", "twJoin", "cva"] : [],
    ignoreTaggedTemplateExpressionIdentifiers: ["weappTwIgnore"],
    patch: {
      filter(className) {
        return !_escape.isAllowedClassName.call(void 0, className);
      }
    }
  };
}



exports.getDefaultOptions = getDefaultOptions;
